### PR TITLE
Megatron work-arounds

### PR DESCRIPTION
**Pull request by borisfom**
# What does this PR do ?

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 1]
```diff
@@ -94,47 +114,28 @@ def __init__(
             use_cpu_initialization=use_cpu_initialization,
         )
 
-        if activation in ['geglu', 'reglu', 'swiglu']:
-            # Separate linear layer for *GLU activations.
+        if activation == 'geglu':
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:117]: Why are we removing the other activations here?

File: nemo/collections/nlp/models/question_answering/qa_model.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def forward(self, input_ids, attention_mask, token_type_ids):
         if isinstance(hidden_states, tuple):
             hidden_states = hidden_states[0]
         logits = self.classifier(hidden_states=hidden_states)
-        return logits
+        return logits.float()
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/question_answering/qa_model.py:67]: Do we really need this conversion to float?

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 3]
```diff
@@ -236,11 +228,11 @@ def __init__(
             )
         else:
             assert attention_type == AttnType.cross_attn
-            self.query = tensor_parallel.ColumnParallelLinear(
+            self.query = tensor_parallel.ColumnLinear(
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:231]: You probably want to remove `tensor_parallel.` here.

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Just marking a few changes here so we don't merge yet.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 4]
```diff
@@ -58,6 +60,24 @@
 """
 
 
+class ColumnLinear(tensor_parallel.ColumnParallelLinear):
```
Comment (ramanathan831) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:63]: if APEX is false, then tensor_parallel import will not happen, resulting in a tensorparallel not found kind of error right?

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 5]
```diff
@@ -94,47 +114,28 @@ def __init__(
             use_cpu_initialization=use_cpu_initialization,
         )
 
-        if activation in ['geglu', 'reglu', 'swiglu']:
```
Comment (ramanathan831) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:97]: any specific reason for taking off reglu and swiglu?

File: nemo/collections/nlp/models/text_classification/text_classification_model.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ def forward(self, input_ids, attention_mask, token_type_ids):
         if isinstance(hidden_states, tuple):
             hidden_states = hidden_states[0]
         logits = self.classifier(hidden_states=hidden_states)
-        return logits
+        return logits.float()
```
Comment (ramanathan831) [inline nemo/collections/nlp/models/text_classification/text_classification_model.py:88]: Should we add for other models too like token_classification?

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 4 above)
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:63]: Why is this needed?

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 5 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:97]: This looks like a merge issue, I will address that.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 4 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:63]: tensor_parallel.ColumnParallelLinear won't export - it has Autograd Function in execution path during inference - which is wrong and I have reported the issue to Apex - but we need some workaround in place now.

File: nemo/collections/nlp/models/text_classification/text_classification_model.py (hunk 6 above)
Comment (borisfom) [reply nemo/collections/nlp/models/text_classification/text_classification_model.py:88]: This is workaround for uncertain output format when using autocast - I am still exploring how to avoid it by forcing FP32 on export side. I guess, it may be needed for some Pytorch scenarios as well. Probably, should be flagged as AMP issue.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 1 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py]: Fixed.

File: nemo/collections/nlp/models/question_answering/qa_model.py (hunk 2 above)
Comment (borisfom) [reply nemo/collections/nlp/models/question_answering/qa_model.py:67]: Explained below.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 3 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py]: Done.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 4 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:54]: APEX is now a requirement.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 5 above)
Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:97]: Done.

File: nemo/collections/nlp/models/text_classification/text_classification_model.py (hunk 6 above)
Comment (borisfom) [reply nemo/collections/nlp/models/text_classification/text_classification_model.py:88]: Added reasonable ones.

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/text_classification/text_classification_model.py:88]: Have you tried setting the precision to 32 within the `.nemo` model config before running export? Megatron models have a precision attribute that is not a part of the trainer and lives inside the model config.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 7]
```diff
@@ -97,9 +108,9 @@ def __init__(
         if activation in ['geglu', 'reglu', 'swiglu']:
             # Separate linear layer for *GLU activations.
             # Source: https://github.com/huggingface/transformers/blob/bee361c6f1f7704f8c688895f2f86f6e5ff84727/src/transformers/models/t5/modeling_t5.py#L292
-            self.dense_h_to_4h_2 = tensor_parallel.ColumnParallelLinear(
+            self.dense_h_to_4h_2 = ColumnLinear(
                 hidden_size,
-                ffn_hidden_size,  # NOTE: When using *glu, divide ffn dim by 2/3 to keep overall params the same.
+                ffn_hidden_size,  # NOTE: When using geglu, divide ffn dim by 2/3 to keep overall params the same.
```
Comment (ramanathan831) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:113]: Replaced during one of the merges, I think this should be for the functions *glu like reglu too, instead of only geglu. Correct me if I am wrong

Comment (borisfom) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:113]: You're right, missed it. Will clean it up.

**Review by ramanathan831 — APPROVED** [review]

**Review by MaximumEntropy — APPROVED** [review]
